### PR TITLE
Better behavior for the default test task when no tests were found

### PR DESCRIPTION
**Pull request by kelvinst**
This born in a discussion on this issue #5556.

- [x] Show a warning instead of an error when the test helper is not found
- [x] Improve the message for when no test path is given
- [x] Test scenarios for each of the above

### CODE REVIEW

**Comment by kelvinst** [comment]
Ready to review, be my guest! 😛

**Comment by josevalim** [comment]
I have added some comments. Instead of creating new fixtures files, it is preferable to use an existing fixture and then delete the test directory, for example.

File: lib/mix/lib/mix/compilers/test.ex [hunk 1]
```diff
@@ -37,7 +37,11 @@ defmodule Mix.Compilers.Test do
         :noop
 
       [] ->
-        Mix.shell.error "Test patterns did not match any file: " <> Enum.join(test_patterns, ", ")
+        if Enum.empty?(test_patterns) do
```
Comment (lexmag) [inline lib/mix/lib/mix/compilers/test.ex]: I think `if test_patterns == [] do` would be even better.

File: lib/mix/lib/mix/tasks/test.ex [hunk 2]
```diff
@@ -362,7 +362,15 @@ defmodule Mix.Tasks.Test do
     if File.exists?(file) do
       Code.require_file file
     else
-      Mix.raise "Cannot run tests because test helper file #{inspect file} does not exist"
```
Comment (kelvinst) [inline lib/mix/lib/mix/tasks/test.ex]: You're totally right! We could stick with the old one. I'll remove the test for this too.

**Comment by kelvinst** [comment]
@josevalim, @lexmag done! Anything else?

File: lib/mix/lib/mix/tasks/test.ex [hunk 3]
```diff
@@ -225,7 +225,7 @@ defmodule Mix.Tasks.Test do
     ex_unit_opts = ex_unit_opts(opts)
     ExUnit.configure(ex_unit_opts)
 
-    test_paths = project[:test_paths] || ["test"]
+    test_paths = project[:test_paths] || default_test_paths
```
Comment (whatyouhide) [inline lib/mix/lib/mix/tasks/test.ex]: You need parentheses when calling `default_test_paths/1`, it emits a warning otherwise that you can see on Travis as well :)

Comment (kelvinst) [inline lib/mix/lib/mix/tasks/test.ex]: Curious, never saw that warning before. Changing!

Comment (kelvinst) [inline lib/mix/lib/mix/tasks/test.ex]: Done!

File: lib/mix/lib/mix/compilers/test.ex [hunk 4]
```diff
@@ -37,7 +37,11 @@ defmodule Mix.Compilers.Test do
         :noop
 
       [] ->
-        Mix.shell.error "Test patterns did not match any file: " <> Enum.join(test_patterns, ", ")
+        if test_patterns == [] do
+          Mix.shell.info "There are no tests to run"
+        else
+          Mix.shell.error "Test patterns did not match any file: " <> Enum.join(test_patterns, ", ")
+        end
```
Comment (whatyouhide) [inline lib/mix/lib/mix/compilers/test.ex]: For consistency with the other clauses in this `case`, we could do this here:

```elixir
[] when test_patterns == [] ->
  Mix.shell.info "There are no tests to run"
  :noop

[] ->
  Mix.shell.error "Test patterns did not match any file: " <> Enum.join(test_patterns, ", ")
  :noop
```

Wdyt? :)

File: lib/mix/lib/mix/tasks/test.ex [hunk 5]
```diff
@@ -122,8 +122,8 @@ defmodule Mix.Tasks.Test do
   ## Configuration
 
     * `:test_paths` - list of paths containing test files, defaults to
-      `["test"]`. It is expected all test paths to contain a `test_helper.exs`
-      file.
+      `["test"]` if the `test` path exists, otherwise, it defaults to `[]`. It
+      is expected all test paths to contain a `test_helper.exs` file.
```
Comment (whatyouhide) [inline lib/mix/lib/mix/tasks/test.ex]: We can rephrase this as:

> [...] defaults to `["test"]` if the `test` directory exists, otherwise it defaults to `[]`. It is expected that all test paths contain a `test_helper.exs` file.

Wdyt?

**Comment by kelvinst** [comment]
Done!

**Comment by whatyouhide** [comment]
Thanks a lot @kelvinst!

**Comment by kelvinst** [comment]
You're welcome!
